File: admin-web/src/lib/api/transfer.ts
/** 代付/转账 API（后台 + 商户端）。对齐 epay transfer + 后端 dto.TransferView。 */
import { request, type PageResult } from './client'

/** 代付记录（对齐后端 dto.TransferView / mock/transfer.ts TransferRecord） */
export interface TransferRecord {
  biz_no: string
  pay_order_no: string
  uid: number // 0=管理员发起
  type: string // alipay/wxpay/qqpay/bank
  channel: number
  account: string
  username: string
  money: string
  costmoney: string
  desc: string
  addtime: string
  paytime: string | null
  status: number // 0处理中 1成功 2失败
  result: string
}

/** 代付概况统计（对齐后端 dto.TransferStats） */
export interface TransferStats {
  total: number
  totalMoney: number
  successMoney: number
  successCount: number
  processingCount: number
  failCount: number
}

export interface TransferListParams {
  page?: number
  pageSize?: number
  keyword?: string // 交易号/收款账号/姓名
  uid?: number
  type?: string
  status?: number
}

/** 发起代付入参（后台/商户共用） */
export interface TransferCreateReq {
  biz_no?: string
  type: string
  channel?: number
  account: string
  username?: string
  money: string
  desc?: string
  password: string // 后台=管理员密码 / 商户=登录密码
}

// ===== 后台 =====
export function fetchTransfers(params: TransferListParams = {}): Promise<PageResult<TransferRecord>> {
  return request<PageResult<TransferRecord>>('/admin/transfers', { query: { ...params } })
}

export function fetchTransferStats(params: TransferListParams = {}): Promise<TransferStats> {
  return request<TransferStats>('/admin/transfers/stats', { query: { ...params } })
}

export function createTransfer(body: TransferCreateReq): Promise<{ biz_no: string }> {
  return request('/admin/transfers', { method: 'POST', body })
}

/** 批量代付单条结果（C-2） */
export interface BatchItemResult {
  index: number
  account: string
  biz_no?: string
  success: boolean
  msg?: string
}

/** 后台批量代付（C-2）：一次校验密码，逐条处理，返回每条结果 */
export function createTransferBatch(
  password: string,
  items: TransferCreateReq[],
): Promise<{ results: BatchItemResult[]; success: number; total: number }> {
  return request('/admin/transfers/batch', { method: 'POST', body: { password, items } })
}

export function setTransferStatus(bizNo: string, status: number, result = ''): Promise<{ biz_no: string; status: number }> {
  return request(`/admin/transfers/${encodeURIComponent(bizNo)}/status`, {
    method: 'PUT',
    body: { status, result },
  })
}

export function refundTransfer(bizNo: string): Promise<{ biz_no: string }> {
  return request(`/admin/transfers/${encodeURIComponent(bizNo)}/refund`, { method: 'POST' })
}

export function deleteTransfer(bizNo: string): Promise<{ biz_no: string }> {
  return request(`/admin/transfers/${encodeURIComponent(bizNo)}`, { method: 'DELETE' })
}

// ===== 商户端 =====
export function fetchMerchantTransfers(params: TransferListParams = {}): Promise<PageResult<TransferRecord>> {
  return request<PageResult<TransferRecord>>('/merchant/transfers', { query: { ...params } })
}

export function createMerchantTransfer(body: TransferCreateReq): Promise<{ biz_no: string }> {
  return request('/merchant/transfer', { method: 'POST', body })
}

File: 0538pay-api/internal/model/transfer.go
package model

import (
	"time"

	"github.com/shopspring/decimal"
)

// Transfer 代付/转账记录。一行 = 一笔对外打款（后台主动发起 或 商户发起）。
// 对齐 epay pre_transfer + 前端 mock/transfer.ts。自研表名 pay_transfer，不复制 epay 表。
//
// 主键为 BizNo（19 位交易号），既是业务号也是幂等键：同一交易号只能落一次（对齐 epay）。
// 资金语义（对齐 epay lib/Transfer）：
//   - 商户发起(UID>0)：落库时即时从余额扣减 CostMoney（含手续费），失败/退回时退回 CostMoney。
//   - 后台发起(UID=0)：CostMoney=Money，不收手续费、不扣款、不涉及退款。
//   - 退款只在 status=0(处理中) 时执行，用条件 UPDATE 防重复退款。
//   - Money=到账金额（打给收款方），CostMoney=商户实际被扣（Money+手续费）。
type Transfer struct {
	BizNo      string          `gorm:"primaryKey;column:biz_no;size:19" json:"biz_no"`   // 交易号（19位数字，幂等键）
	PayOrderNo string          `gorm:"size:80" json:"pay_order_no"`                      // 第三方转账单号（渠道返回）
	UID        uint            `gorm:"index;not null" json:"uid"`                        // 商户号（0=后台管理员发起）
	Type       string          `gorm:"size:10;not null" json:"type"`                     // 付款方式 alipay/wxpay/qqpay/bank
	Channel    int             `gorm:"not null;default:0" json:"channel"`                // 使用的支付通道 id
	Account    string          `gorm:"size:128;not null" json:"account"`                 // 收款账号
	Username   string          `gorm:"size:128" json:"username"`                         // 收款人姓名（选填，留空不校验）
	Money      decimal.Decimal `gorm:"type:decimal(18,4);not null" json:"-"`             // 到账金额（原始）
	CostMoney  decimal.Decimal `gorm:"type:decimal(18,4)" json:"-"`                      // 商户扣款=到账+手续费（原始）
	AddTime    time.Time       `gorm:"index" json:"-"`                                   // 提交时间（原始）
	PayTime    *time.Time      `json:"-"`                                                // 付款成功时间（status=1 时写入，原始）
	Status     int8            `gorm:"not null;default:0;index" json:"status"`           // 0处理中 1成功 2失败
	Desc       string          `gorm:"size:80" json:"desc"`                              // 转账备注（≤32字）
	Result     string          `gorm:"size:255" json:"result"`                           // 结果消息（失败原因/撤销说明）
	Ext        string          `gorm:"type:text" json:"-"`                               // 扩展（微信确认收款 jumpurl 等）
	API        int8            `gorm:"column:api;not null;default:0" json:"-"`           // 发起来源 0后台 1API（E-6，对齐 epay pre_transfer.api）
}

func (Transfer) TableName() string { return "pay_transfer" }
